File: src/parser.zig
const mem = @import("mem");
const util = @import("util");
const collections = @import("collections");

const generator = @import("generator/mod.zig");
const checker = @import("checker/mod.zig");
const node = @import("node/mod.zig");
const lexer = @import("lexer/mod.zig");
const token = lexer.token;

const ARGUMENT_MAX: u32 = 20;

const Precedence = enum(usize) {
    Nothing,
    Assignment,
    Equality,
    Comparison,
    Term,
    Factor,
    Unary,
    Call,
    Declaration,
    Scope,
};

const Fn = *const fn (parser: *Parser, lexer: *lexer.Lexer) void;
const Ruler = struct {
    callbacks: collections.Array(Rule),
    transforms: collections.Array(I.Transform),

    const I = Iter(token.Token, Rule);

    fn new(allocator: *mem.Arena) error{OutOfMemory, OutOfBounds}!Ruler {
        const zone = util.tracy.initZone(@src(), .{.name = "Ruler::new"});
        defer zone.deinit();

        var self: Ruler = undefined;
        const Ts = [_]type {token.Keyword, token.Symbol, token.Operator};
        const token_kind_len = @typeInfo(token.TokenKind).@"enum".fields.len;
        comptime var len = token_kind_len - Ts.len;

        inline for (Ts) |T| {
            len += @typeInfo(T).@"enum".fields.len;
        }

        const transforms: [token_kind_len]I.Transform = .{
            token.identity_int, token.identity_int, token.identity_int, token.identity_int, token.keyword_int, token.symbol_int, token.operator_int,
        };

        self.callbacks = try I.new(
            &[_]I.Pair {
                I.Pair { .t = token.Token.IDEN, .f = Rule.new(Parser.identifier, null, .Nothing), },
                I.Pair { .t = token.Token.NUMBER, .f = Rule.new(Parser.number, null, .Nothing), },
                I.Pair { .t = token.Token.LET, .f = Rule.new(Parser.let, null, .Declaration), },
                I.Pair { .t = token.Token.PARENTESISLEFT, .f = Rule.new(Parser.group, Parser.call, .Call), },
                I.Pair { .t = token.Token.BRACELEFT, .f = Rule.new(Parser.scope, Parser.construct, .Assignment), },
                I.Pair { .t = token.Token.DOT, .f = Rule.new(null, Parser.property, .Call) },
                I.Pair { .t = token.Token.MINUS, .f = Rule.new(Parser.unary, Parser.binary, .Term), },
                I.Pair { .t = token.Token.PLUS, .f = Rule.new(null, Parser.binary, .Term), },
                I.Pair { .t = token.Token.BANG, .f = Rule.new(Parser.unary, null, .Equality), },
                I.Pair { .t = token.Token.BANGEQUAL, .f = Rule.new(null, Parser.binary, .Equality), },
                I.Pair { .t = token.Token.EQUALEQUAL, .f = Rule.new(null, Parser.binary, .Equality), },
                I.Pair { .t = token.Token.STAR, .f = Rule.new(null, Parser.binary, .Factor), },
                I.Pair { .t = token.Token.SLASH, .f = Rule.new(null, Parser.binary, .Factor), },
                I.Pair { .t = token.Token.GREATER, .f = Rule.new(null, Parser.binary, .Comparison), },
                I.Pair { .t = token.Token.LESS, .f = Rule.new(null, Parser.binary, .Comparison), },
                I.Pair { .t = token.Token.GREATEREQUAL, .f = Rule.new(null, Parser.binary, .Comparison), },
                I.Pair { .t = token.Token.LESSEQUAL, .f = Rule.new(null, Parser.binary, .Comparison), },
                I.Pair { .t = token.Token.STRING, .f = Rule.new(Parser.string, null, .Nothing), },
            },
            &transforms,
            len,
            Rule.new(null, null, .Nothing),
            allocator,
        );

        errdefer self.callbacks.deinit(allocator);

        self.transforms = try collections.Array(I.Transform).new(transforms.len, allocator);

        for (0..transforms.len) |i| {
            self.transforms.items[i] = transforms[i];
        }

        return self;
    }

    fn from_token(self: *const Ruler, t: token.Token) Rule {
        const zone = util.tracy.initZone(@src(), .{.name = "Ruler::from_token"});
        defer zone.deinit();

        const index = @intFromEnum(t);
        const pos = self.transforms.items[index](t, index);

        return self.callbacks.items[pos];
    }

    fn deinit(self: *Ruler, allocator: *mem.Arena) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Ruler::deinit"});
        defer zone.deinit();

        self.transforms.deinit(allocator);
        self.callbacks.deinit(allocator);
    }
};

const Rule = struct {
    prefix: ?Fn,
    infix: ?Fn,
    precedence: Precedence,

    fn new(prefix: ?Fn, infix: ?Fn, precedence: Precedence) Rule {
        const zone = util.tracy.initZone(@src(), .{.name = "Rule::new"});
        defer zone.deinit();

        return Rule{
            .prefix = prefix,
            .infix = infix,
            .precedence = precedence,
        };
    }
};

fn Iter(T: type, F: type) type {
    return struct {
        const Self = @This();

        const Pair = struct {
            t: T,
            f: F,
        };

        const Transform = *const fn (T, i: usize) usize;

        fn new(
            pairs: []const Pair,
            transforms: []const Transform,
            len: u32,
            zero: F,
            allocator: *mem.Arena,
        ) error{OutOfMemory, OutOfBounds}!collections.Array(F) {
            const zone = util.tracy.initZone(@src(), .{.name = "Iter::new"});
            defer zone.deinit();

            var callbacks = try collections.Array(F).new(len, allocator);
            errdefer callbacks.deinit(allocator);

            for (0..len) |i| {
                try callbacks.set(zero, i);
            }

            for (pairs) |pair| {
                const i = @intFromEnum(pair.t);
                const pos = transforms[i](pair.t, i);

                try callbacks.set(pair.f, pos);
            }

            return callbacks;
        }
    };
}

pub const Parser = struct {
    ruler: Ruler,
    manager: node.Tree,
    arena: *mem.Arena,

    pub fn new(allocator: *mem.Arena) error{ OutOfBounds, OutOfMemory }!Parser {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::new"});
        defer zone.deinit();

        var self = Parser{
            .ruler = undefined,
            .manager = undefined,
            .arena = try allocator.child("Parser", mem.PAGE_SIZE),
        };

        errdefer self.arena.deinit();

        self.manager = try node.Tree.new(self.arena);
        errdefer self.manager.deinit(self.arena);

        self.ruler = try Ruler.new(self.arena);
        errdefer self.ruler.deinit(self.arena);

        return self;
    }

    pub fn next(self: *Parser, tokenizer: *lexer.Lexer) ?*node.Tree {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::next"});
        defer zone.deinit();

        self.manager.clear();

        _ = tokenizer.next() orelse return null;

        switch (tokenizer.previous) {
            .Keyword => |keyword| switch (keyword) {
                .proc => self.procedure(tokenizer),
                .@"type" => self.typ(tokenizer),
                else => @panic("TODO: No other statement kind is accepted for now"),
            },
            else => @panic("TODO: do not accept expressions here for now"),
        }

        return &self.manager;
    }

    fn group(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::group"});
        defer zone.deinit();

        self.parse(.Assignment, tokenizer);
        if (!tokenizer.match(token.Token.PARENTESISRIGHT)) @panic("TODO");
    }

    fn unary(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::unary"});
        defer zone.deinit();

        const operator = tokenizer.previous.Operator;
        self.parse(.Unary, tokenizer);

        switch (operator) {
            .Bang => self.manager.nodes.push(node.Node{ .Unary = node.Unary.new(.Bang) }) catch @panic("TODO"),
            .Minus => self.manager.nodes.push(node.Node{ .Unary = node.Unary.new(.Negate) }) catch @panic("TODO"),
            else => @panic("TODO"),
        }
    }

    fn binary(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::binary"});
        defer zone.deinit();

        const operator = tokenizer.previous.Operator;

        self.parse(
            @enumFromInt(
                @intFromEnum(
                    self.ruler.from_token(tokenizer.previous).precedence,
                ) + 1,
            ),
            tokenizer,
        );

        switch (operator) {
            .BangEqual => {
                self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Eq) }) catch @panic("TODO");
                self.manager.nodes.push(node.Node{ .Unary = node.Unary.new(.Bang) }) catch @panic("TODO");
            },
            .GreaterEqual => {
                self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Lt) }) catch @panic("TODO");
                self.manager.nodes.push(node.Node{ .Unary = node.Unary.new(.Bang) }) catch @panic("TODO");
            },
            .LessEqual => {
                self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Gt) }) catch @panic("TODO");
                self.manager.nodes.push(node.Node{ .Unary = node.Unary.new(.Bang) }) catch @panic("TODO");
            },
            .Minus => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Add) }) catch @panic("TODO"),
            .EqualEqual => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Eq) }) catch @panic("TODO"),
            .Greater => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Gt) }) catch @panic("TODO"),
            .Less => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Lt) }) catch @panic("TODO"),
            .Plus => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Add) }) catch @panic("TODO"),
            .Star => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Mul) }) catch @panic("TODO"),
            .Slash => self.manager.nodes.push(node.Node{ .Binary = node.Binary.new(.Div) }) catch @panic("TODO"),

            else => @panic("TODO"),
        }
    }

    fn identifier(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::identifier"});
        defer zone.deinit();

        self.manager.nodes.push(node.Node{
            .Identifier = node.Identifier.new(tokenizer.previous.Identifier),
        }) catch @panic("TODO");
    }

    fn number(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::number"});
        defer zone.deinit();

        self.manager.nodes.push(node.Node{ .Number = node.Number.new(
            tokenizer.previous.Number,
        ) }) catch @panic("TODO");
    }

    fn construct(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::construct"});
        defer zone.deinit();

        const index = self.manager.construct_values.start() catch @panic("TODO");

        while (!tokenizer.match(token.Token.BRACERIGHT)) {
            if (!tokenizer.match(token.Token.COMMA) and !tokenizer.previous.eql(token.Token.BRACELEFT)) {
                @panic("Do not forget to separete arguments with comma");
            }

            const name = tokenizer.current;

            if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
            if (!tokenizer.match(token.Token.DOUBLECOLON)) @panic("TODO");

            self.parse(.Assignment, tokenizer);

            self.manager.construct_values.extend(node.Construct.Value.new(
                name.Identifier,
                @intCast(self.manager.nodes.len - 1),
            )) catch @panic("TODO");
        }

        self.manager.nodes.push(node.Node{
            .Construct = node.Construct.new(index),
        }) catch @panic("TODO");
    }

    fn call(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::call"});
        defer zone.deinit();

        var len: util.Index = 0;

        while (!tokenizer.match(token.Token.PARENTESISRIGHT)) {
            if (!tokenizer.match(token.Token.COMMA) and !tokenizer.previous.eql(token.Token.PARENTESISLEFT)) {
                @panic("TODO: Show that user cannot place a comma at the start of the function call");
            }

            self.parse(.Assignment, tokenizer);
            len += 1;
        }

        self.manager.nodes.push(node.Node{
            .Call = node.Call.new(len),
        }) catch @panic("TODO");
    }

    fn typ(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::typ"});
        defer zone.deinit();

        const name = tokenizer.current;

        if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");

        if (tokenizer.match(token.Token.BRACELEFT)) {
            while (!tokenizer.match(token.Token.BRACERIGHT)) {
                const field_name = tokenizer.current;

                if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
                if (!tokenizer.match(token.Token.DOUBLECOLON)) @panic("TODO");

                const field_type = tokenizer.current;

                if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
                if (!tokenizer.match(token.Token.COMMA)) @panic("TODO");

                self.manager.type_fields.push(node.Type.Field.new(
                    field_name.Identifier,
                    field_type.Identifier,
                )) catch @panic("TODO");
            }
        } else {
            if (!tokenizer.match(token.Token.EQUAL)) @panic("TODO");
            const type_size = tokenizer.current;

            if (!tokenizer.match(token.Token.NUMBER)) @panic("TODO");
            if (!tokenizer.match(token.Token.SEMICOLON)) @panic("TODO");

            self.manager.nodes.push(node.Node{ .Number = node.Number.new(
                type_size.Number,
            ) }) catch @panic("TODO");
        }

        self.manager.nodes.push(node.Node{ .Type = node.Type.new(
            name.Identifier,
        ) }) catch @panic("TODO");
    }

    fn property(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::property"});
        defer zone.deinit();

        const name = tokenizer.current;

        if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");

        self.manager.nodes.push(node.Node{
            .Property = node.Property.new(name.Identifier),
        }) catch @panic("TODO");
    }

    fn procedure(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::procedure"});
        defer zone.deinit();

        const name = tokenizer.current;

        if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
        if (!tokenizer.match(token.Token.PARENTESISLEFT)) @panic("TODO");

        while (!tokenizer.match(token.Token.PARENTESISRIGHT)) {
            if (!tokenizer.match(token.Token.COMMA) and !tokenizer.previous.eql(token.Token.PARENTESISLEFT)) {
                @panic("TODO: Parameters should start with parentesis and the first element is a identifier");
            }

            const param = tokenizer.current;

            if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
            if (!tokenizer.match(token.Token.DOUBLECOLON)) @panic("TODO");

            const kind = tokenizer.current;

            if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");

            self.manager.parameters.push(node.Procedure.Parameter.new(
                param.Identifier,
                kind.Identifier,
            )) catch @panic("TODO");
        }

        if (!tokenizer.match(token.Token.DOUBLECOLON)) @panic("TODO");

        self.parse(.Call, tokenizer);

        if (!tokenizer.match(token.Token.BRACELEFT)) @panic("TODO");

        self.scope(tokenizer);

        self.manager.nodes.push(node.Node{ .Procedure = node.Procedure.new(
            name.Identifier,
        ) }) catch @panic("TODO");
    }

    fn let(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::let"});
        defer zone.deinit();

        const mutable = tokenizer.match(token.Token.MUT);
        const iden = tokenizer.current;
        _ = mutable;

        if (!tokenizer.match(token.Token.IDEN)) @panic("TODO");
        if (!tokenizer.match(token.Token.DOUBLECOLON)) @panic("TODO");

        self.parse(.Assignment, tokenizer);

        if (!tokenizer.match(token.Token.EQUAL)) @panic("TODO");

        self.parse(.Assignment, tokenizer);

        if (!tokenizer.match(token.Token.SEMICOLON)) @panic("TODO");

        self.manager.nodes.push(node.Node{ .Let = node.Let.new(
            iden.Identifier,
        ) }) catch @panic("TODO");
    }

    fn string(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::string"});
        defer zone.deinit();

        _ = tokenizer;
        _ = self;
        @panic("TODO");
    }

    fn scope(self: *Parser, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::scope"});
        defer zone.deinit();

        var len: util.Index = 0;

        while (!tokenizer.match(token.Token.BRACERIGHT)) {
            self.parse(.Assignment, tokenizer);

            len += 1;
        }

        self.manager.nodes.push(node.Node {.Scope = node.Scope.new(len)}) catch @panic("TODO");
    }

    fn parse(self: *Parser, precedence: Precedence, tokenizer: *lexer.Lexer) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::parse"});
        defer zone.deinit();

        tokenizer.advance();

        if (self.ruler.from_token(tokenizer.previous).prefix) |prefix| {
            prefix(self, tokenizer);
        } else {
            @panic("TODO: Show error");
        }

        var r = self.ruler.from_token(tokenizer.current);

        while (@intFromEnum(precedence) <= @intFromEnum(r.precedence)) {
            if (r.infix) |infix| {
                _ = tokenizer.next() orelse @panic("TODO");
                infix(self, tokenizer);
            } else {
                break;
            }

            r = self.ruler.from_token(tokenizer.current);
        }
    }

    pub fn deinit(self: *Parser) void {
        const zone = util.tracy.initZone(@src(), .{.name = "Parser::deinit"});
        defer zone.deinit();

        self.ruler.deinit(self.arena);
        self.manager.deinit(self.arena);
        self.arena.deinit();
    }
};

File: compiler/src/parser.rs
use mem::Arena;

use crate::checker::{BinaryOperator, TypeChecker, UnaryOperator};
use crate::lexer::{Keyword, Lexer, Operator, Symbol, Token};

#[repr(u32)]
#[derive(Clone, Copy)]
enum Precedence {
    Nothing,
    Assignment,
    Or,
    And,
    Equality,
    Comparison,
    Term,
    Factor,
    Unary,
    Call,
    Primary,
}

type Fn = fn(&mut Parser);
struct Rule {
    prefix: Option<Fn>,
    infix: Option<Fn>,
    precedence: Precedence,
}

pub struct Parser {
    previous: Token,
    current: Token,
    lexer: Lexer,
    checker: TypeChecker,
}

impl Precedence {
    fn from_integer(integer: u32) -> Precedence {
        unsafe { std::mem::transmute(integer) }
    }

    fn to_int(&self) -> u32 {
        *self as u32
    }

    fn inc(&self) -> Precedence {
        Precedence::from_integer(self.to_int() + 1)
    }
}

impl Rule {
    fn new(prefix: Option<Fn>, infix: Option<Fn>, precedence: Precedence) -> Rule {
        Rule {
            prefix,
            infix,
            precedence,
        }
    }

    fn from_token(token: Token) -> Rule {
        match token {
            Token::Symbol(symbol) => {
                if let Symbol::ParentesisLeft = symbol {
                    Rule::new(Some(grouping), None, Precedence::Assignment)
                } else {
                    Rule::new(None, None, Precedence::Nothing)
                }
            }
            Token::Operator(operator) => match operator {
                Operator::Minus => Rule::new(Some(unary), Some(binary), Precedence::Term),
                Operator::Plus => Rule::new(None, Some(binary), Precedence::Term),
                Operator::Bang => Rule::new(Some(unary), None, Precedence::Equality),
                Operator::BangEqual | Operator::EqualEqual => {
                    Rule::new(None, Some(binary), Precedence::Equality)
                }
                Operator::Slash | Operator::Star => {
                    Rule::new(None, Some(binary), Precedence::Factor)
                }
                Operator::Greater
                | Operator::GreaterEqual
                | Operator::Less
                | Operator::LessEqual => Rule::new(None, Some(binary), Precedence::Comparison),
            },
            Token::Keyword(keyword) => match keyword {
                Keyword::False | Keyword::True => {
                    Rule::new(Some(literal), None, Precedence::Nothing)
                }
                _ => Rule::new(None, None, Precedence::Nothing),
            },
            Token::Number(_) => Rule::new(Some(number), None, Precedence::Nothing),
            Token::Identifier(_) => Rule::new(Some(identifier), None, Precedence::Nothing),
            Token::String(_) => Rule::new(Some(string), None, Precedence::Nothing),
            _ => Rule::new(None, None, Precedence::Nothing),
        }
    }
}

impl Parser {
    pub fn new(input: &str, output: &str, arena: &mut Arena) -> Parser {
        let mut lexer = Lexer::new(input, arena);
        let mut checker = TypeChecker::new(output, arena);

        checker.start_scope();

        Parser {
            previous: Token::Eof,
            current: lexer.next(),
            lexer,
            checker,
        }
    }

    fn advance(&mut self) {
        self.previous = self.current;
        self.current = self.lexer.next();
    }

    fn consume(&mut self, token: Token) {
        if self.current == token {
            self.advance();
        } else {
            panic!("Failed to consume token");
        }
    }

    fn assert(&mut self, token: Token) -> bool {
        if self.current == token {
            self.advance();
            true
        } else {
            false
        }
    }

    fn parse(&mut self, precedence: Precedence) {
        self.advance();

        if let Some(f) = Rule::from_token(self.previous).prefix {
            f(self);
        } else {
            panic!("should not happen");
        }

        let mut rule = Rule::from_token(self.current);

        while precedence.to_int() <= rule.precedence.to_int() {
            if let Some(f) = rule.infix {
                self.advance();
                f(self);
            } else {
                break;
            }

            rule = Rule::from_token(self.current);
        }
    }

    pub fn next(&mut self) -> bool {
        if !self.assert(Token::EOF) {
            statement(self);
            true
        } else {
            false
        }
    }

    pub fn deinit(&mut self) {
        self.checker.deinit(&self.lexer.words);
    }
}

fn statement(parser: &mut Parser) {
    match parser.current {
        Token::Keyword(keyword) => match keyword {
            Keyword::Let => variable(parser),
            Keyword::Procedure => procedure(parser),
            Keyword::Type => typ(parser),
            Keyword::Case => case(parser),
            _ => {}
        },
        _ => expression(parser),
    }
}

fn expression(parser: &mut Parser) {
    parser.parse(Precedence::Assignment);
}

fn grouping(parser: &mut Parser) {
    expression(parser);
    parser.consume(Token::PARENTESISRIGHT);
}

fn unary(parser: &mut Parser) {
    let Token::Operator(operator) = parser.previous else {
        panic!("Should not happen")
    };

    parser.parse(Precedence::Unary);

    match operator {
        Operator::Bang => parser.checker.push_unary(UnaryOperator::Negation),
        Operator::Minus => parser.checker.push_unary(UnaryOperator::Oposite),
        _ => {}
    }
}

fn binary(parser: &mut Parser) {
    let rule = Rule::from_token(parser.previous);
    let Token::Operator(operator) = parser.previous else {
        panic!("Should not happen")
    };

    parser.parse(rule.precedence.inc());

    match operator {
        Operator::BangEqual => {
            parser
                .checker
                .push_binary(BinaryOperator::Eq, &parser.lexer.words);
            parser.checker.push_unary(UnaryOperator::Negation);
        }
        Operator::GreaterEqual => {
            parser
                .checker
                .push_binary(BinaryOperator::Lt, &parser.lexer.words);
            parser.checker.push_unary(UnaryOperator::Negation);
        }
        Operator::LessEqual => {
            parser
                .checker
                .push_binary(BinaryOperator::Gt, &parser.lexer.words);
            parser.checker.push_unary(UnaryOperator::Negation);
        }
        Operator::EqualEqual => parser
            .checker
            .push_binary(BinaryOperator::Eq, &parser.lexer.words),
        Operator::Greater => parser
            .checker
            .push_binary(BinaryOperator::Gt, &parser.lexer.words),
        Operator::Less => parser
            .checker
            .push_binary(BinaryOperator::Lt, &parser.lexer.words),
        Operator::Plus => parser
            .checker
            .push_binary(BinaryOperator::Add, &parser.lexer.words),
        Operator::Minus => parser
            .checker
            .push_binary(BinaryOperator::Sub, &parser.lexer.words),
        Operator::Star => parser
            .checker
            .push_binary(BinaryOperator::Mul, &parser.lexer.words),
        Operator::Slash => parser
            .checker
            .push_binary(BinaryOperator::Div, &parser.lexer.words),
        _ => {}
    }
}

fn literal(parser: &mut Parser) {
    let keyword = parser.previous.keyword();

    match keyword {
        Keyword::False => parser.checker.push_boolean(false, &parser.lexer.words),
        Keyword::True => parser.checker.push_boolean(true, &parser.lexer.words),
        _ => {}
    }
}

fn identifier(parser: &mut Parser) {
    let Token::Identifier(range) = parser.previous else {
        panic!("Should not happen")
    };

    if parser.assert(Token::PARENTESISLEFT) {
        call(parser);
        parser.checker.push_range(range);
        parser.checker.push_call(&parser.lexer.words);
    } else if parser.assert(Token::BRACELEFT) {
        construct(parser);
        parser.checker.push_range(range);
        parser.checker.push_construct(&parser.lexer.words);
    } else {
        parser.checker.push_identifier(range, &parser.lexer.words);
    }
}

fn construct(parser: &mut Parser) {
    while !parser.assert(Token::BRACERIGHT) {
        let Token::Identifier(range) = parser.current else {
            panic!("Should not happen");
        };

        parser.advance();
        parser.consume(Token::DOUBLECOLON);

        expression(parser);

        parser.consume(Token::COMMA);
        parser.checker.push_range(range);
    }
}

fn number(parser: &mut Parser) {
    let Token::Number(range) = parser.previous else {
        panic!("Should not happen")
    };

    parser.checker.push_number(range, &parser.lexer.words);
}

fn procedure(parser: &mut Parser) {
    parser.advance();

    let Token::Identifier(iden) = parser.current else {
        panic!("Should not happen")
    };

    parser.advance();
    parser.consume(Token::PARENTESISLEFT);

    parser.checker.start_scope();

    let mut parameter_count: usize = 0;

    while !parser.assert(Token::PARENTESISRIGHT) {
        let Token::Identifier(param) = parser.current else {
            panic!("Should not happen")
        };

        parser.advance();
        parser.consume(Token::DOUBLECOLON);

        let Token::Identifier(kind) = parser.current else {
            panic!("Should not happen")
        };

        parser.advance();
        parser.checker.push_range(param);
        parser.checker.push_range(kind);

        _ = parser.assert(Token::COMMA);

        parser.checker.push_parameter(&parser.lexer.words);
        parameter_count += 1;
    }

    parser.consume(Token::DOUBLECOLON);

    let Token::Identifier(ret) = parser.current else {
        panic!("Should not happen")
    };

    parser.advance();
    parser.consume(Token::BRACELEFT);

    scope(parser);

    parser.checker.push_range(iden);
    parser.checker.push_range(ret);

    parser
        .checker
        .push_procedure(parameter_count, &parser.lexer.words);
}

fn variable(parser: &mut Parser) {
    parser.advance();

    let mutable = parser.assert(Token::MUT);
    let Token::Identifier(iden) = parser.current else {
        panic!("Should not happen")
    };

    parser.advance();
    parser.consume(Token::DOUBLECOLON);

    let Token::Identifier(kind) = parser.current else {
        panic!("Should not happen")
    };

    parser.advance();
    parser.consume(Token::EQUAL);

    expression(parser);

    parser.consume(Token::SEMICOLON);

    parser.checker.push_range(iden);
    parser.checker.push_range(kind);
    parser.checker.push_let(&parser.lexer.words);
}

fn call(parser: &mut Parser) {
    while !parser.assert(Token::PARENTESISRIGHT) {
        _ = parser.assert(Token::COMMA);

        expression(parser);
    }
}

fn case(parser: &mut Parser) {
    parser.advance();

    expression(parser);

    parser.consume(Token::BRACELEFT);

    let mut of_count: usize = 0;
    while !parser.assert(Token::BRACERIGHT) {
        parser.consume(Token::OF);

        match parser.current {
            Token::Identifier(r) => parser.checker.push_identifier(r, &parser.lexer.words),
            Token::Number(r) => parser.checker.push_number(r, &parser.lexer.words),
            Token::Keyword(k) => match k {
                Keyword::True => parser.checker.push_boolean(true, &parser.lexer.words),
                Keyword::False => parser.checker.push_boolean(false, &parser.lexer.words),
                _ => panic!("Expected value, found keyword"),
            },
            _ => panic!("Matcher just accept identifiers, numbers, or booleans"),
        }

        parser.advance();
        parser.consume(Token::ARROW);

        if parser.assert(Token::BRACELEFT) {
            parser.checker.start_scope();
            scope(parser);
        } else {
            expression(parser);
        }

        parser.consume(Token::COMMA);
        of_count += 1;
    }

    parser.checker.push_case(of_count);
}

fn typ(parser: &mut Parser) {
    parser.advance();

    let Token::Identifier(iden) = parser.current else {
        panic!("Should not happen");
    };

    parser.advance();

    let mut field_count: usize = 0;
    let mut size: usize = 0;

    if parser.assert(Token::BRACELEFT) {
        while !parser.assert(Token::BRACERIGHT) {
            let Token::Identifier(field_name) = parser.current else {
                panic!("Should not happen");
            };

            parser.advance();
            parser.consume(Token::DOUBLECOLON);

            let Token::Identifier(field_type) = parser.current else {
                panic!("Should not happen");
            };

            parser.advance();
            parser.consume(Token::COMMA);

            parser.checker.push_range(field_name);
            parser.checker.push_range(field_type);

            field_count += 1;
        }
    } else {
        let Token::Number(size_range) = parser.current else {
            panic!("Should not happen");
        };

        parser.advance();
        parser.consume(Token::SEMICOLON);

        size = util::parse_string(parser.lexer.words.range(size_range));
    }

    parser.checker.push_range(iden);
    parser
        .checker
        .push_type(field_count, size, &parser.lexer.words);
}

fn scope(parser: &mut Parser) {
    while !parser.assert(Token::BRACERIGHT) {
        statement(parser);
    }

    parser.checker.end_scope();
}

fn string(parser: &mut Parser) {
    parser.advance();
}
